Optimize alias_analysis node lookup

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54115 Optimize alias_analysis node lookup**

By using this optimization, we were able to reduce the one round of tracing time for some internal models from 20 mins to under 1 min now. Including the extra time for check tracing, we can now trace models that take one hour to trace to within couple minutes. 

Differential Revision: [D27104047](https://our.internmc.facebook.com/intern/diff/D27104047)

tugsbayasgalan committed:
Optimize alias_analysis node lookup

[ghstack-poisoned]

gmagogsfm commented on torch/csrc/jit/ir/alias_analysis.cpp:
```diff
@@ -1406,8 +1407,9 @@ class AliasDb::WorkingSet {
-    return std::any_of(nodes_.begin(), nodes_.end(), [&](Node* node) {
-      return users.count(node) != 0;
+
+    return std::any_of(users.begin(), users.end(), [&](Node* node) {
```
Nit: rename argument in lambda to `user` to reflect that it is actually part of `users`

tugsbayasgalan replied:
done

gmagogsfm reviewed (approved):
LGTM.

Could you include some data on how much performance improvement we are seeing?

tugsbayasgalan committed:
Update on "Optimize alias_analysis node lookup"


By using this optimization, we were able to reduce the tracing time for some internal models from 20 mins to under 1 min now. 

Differential Revision: [D27104047](https://our.internmc.facebook.com/intern/diff/D27104047)

[ghstack-poisoned]

eellison commented on torch/csrc/jit/ir/alias_analysis.cpp:
```diff
@@ -1452,6 +1453,10 @@ class AliasDb::WorkingSet {
 
   const AliasDb& aliasDb_;
   std::vector<Node*> nodes_;
+  // Extra data structure for nodes for faster look up
```
Can you describe what the Node is being mapped to (what it represents) instead of how it is being used ?

eellison commented on torch/csrc/jit/ir/alias_analysis.cpp:
```diff
@@ -1327,6 +1327,7 @@ class AliasDb::WorkingSet {
   // Add `n` to the working set
   void add(Node* n) {
     nodes_.push_back(n);
+    node_to_index_[n] = nodes_.size() - 1;
```
This isn't quite right: here, you're setting `node_to_index_[n]` to `nodes_.size()`. `nodes_.size()` will get reconstructed each time we try `tryMove`. So you will have multiple unrelated nodes that have the same index value.

eellison replied:
thought node_to_index was a member of aliasDb, but it's amember of working set so we're not duplicating. my bad

eellison reviewed (changes requested):
Logic isn't exactly right, you're setting multiple unrelated nodes to the same index value

eellison reviewed (approved):
JK, looks good!
